datadog/resource_datadog_user.go

```diff
@@ -90,75 +118,244 @@ func buildDatadogUserStruct(d *schema.ResourceData) *datadog.User {
+		authV2 := providerConf.AuthV2
+
+		userRequest := buildDatadogUserV2UpdateStruct(d, d.Id())
+
+		_, _, err := datadogClientV2.UsersApi.UpdateUser(authV2, d.Id()).Body(*userRequest).Execute()
+		if err != nil {
+			return translateClientError(err, "error updating user")
+		}
+		if d.HasChange("roles") {
+			oldRolesI, newRolesI := d.GetChange("roles")
+			oldRoles := oldRolesI.(*schema.Set)
+			newRoles := newRolesI.(*schema.Set)
+			rolesToRemove := oldRoles.Difference(newRoles)
+			rolesToAdd := newRoles.Difference(oldRoles)
+			for _, roleI := range rolesToRemove.List() {
+				role := roleI.(string)
+				userRelation := datadogV2.NewRelationshipToUserWithDefaults()
+				userRelationData := datadogV2.NewRelationshipToUserDataWithDefaults()
+				userRelationData.SetId(d.Id())
+				userRelation.SetData(*userRelationData)
+				_, _, err := datadogClientV2.RolesApi.RemoveUserFromRole(authV2, role).Body(*userRelation).Execute()
+				if err != nil {
+					return translateClientError(err, "error removing user from role")
+				}
+			}
+			for _, roleI := range rolesToAdd.List() {
+				role := roleI.(string)
+				roleRelation := datadogV2.NewRelationshipToUserWithDefaults()
+				roleRelationData := datadogV2.NewRelationshipToUserDataWithDefaults()
+				roleRelationData.SetId(d.Id())
+				roleRelation.SetData(*roleRelationData)
+				_, _, err := datadogClientV2.RolesApi.AddUserToRole(authV2, role).Body(*roleRelation).Execute()
+				if err != nil {
+					return translateClientError(err, "error adding user to role")
+				}
+			}
+		}
+	} else {
+		client := providerConf.CommunityClient
+
+		u := buildDatadogUserStruct(d)
+		u.SetHandle(d.Id())
 
-	if err := client.UpdateUser(*u); err != nil {
-		return translateClientError(err, "error updating user")
+		if err := client.UpdateUser(*u); err != nil {
+			return translateClientError(err, "error updating user")
+		}
 	}
 
 	return resourceDatadogUserRead(d, meta)
 }
 
 func resourceDatadogUserDelete(d *schema.ResourceData, meta interface{}) error {
 	providerConf := meta.(*ProviderConfiguration)
-	client := providerConf.CommunityClient
 
-	// Datadog does not actually delete users, but instead marks them as disabled.
-	// Bypass DeleteUser if GetUser returns User.Disabled == true, otherwise it will 400.
-	if u, err := client.GetUser(d.Id()); err == nil && u.GetDisabled() {
-		return nil
-	}
+	if isV2User(d.Id()) {
+		datadogClientV2 := providerConf.DatadogClientV2
+		authV2 := providerConf.AuthV2
 
-	if err := client.DeleteUser(d.Id()); err != nil {
-		return translateClientError(err, "error deleting user")
+		if httpResponse, err := datadogClientV2.UsersApi.DisableUser(authV2, d.Id()).Execute(); err != nil {
+			if httpResponse != nil && httpResponse.StatusCode == 404 {
+				return nil
+			}
+			return translateClientError(err, "error disabling user")
+		}
+	} else {
+		client := providerConf.CommunityClient
+
+		// Datadog does not actually delete users, but instead marks them as disabled.
```

Comment:
Does v2 not have the same 400 behavior if try to disable an already disabled user?

Reply:
No we get a 404 in this case.